### PR TITLE
rpcdaemon: extract block cache for chain storage access

### PR DESCRIPTION
**Pull request by lupin012**

### CODE REVIEW

File: silkworm/silkrpc/core/cached_chain.cpp [hunk 1]
```diff
@@ -36,6 +36,32 @@ awaitable<std::shared_ptr<BlockWithHash>> read_block_by_number(BlockCache& cache
     co_return block_with_hash;
 }
 
+awaitable<std::shared_ptr<BlockWithHash>> read_block_by_number(BlockCache& cache, const std::shared_ptr<ChainStorage> storage, const rawdb::DatabaseReader& reader, uint64_t block_number) {
+    evmc::bytes32 block_hash;
+    try {
+        block_hash = co_await rawdb::read_canonical_block_hash(reader, block_number);
+    } catch (const std::invalid_argument& iv) {
+        co_return nullptr;
+    }
```
Comment (canepat) [inline silkworm/silkrpc/core/cached_chain.cpp:45]: As discussed, we should still rely on `std::invalid_argument` raised by low-level functions here without catching it at this level. Changing this behaviour consistently in the internal APIs is IMHO has broader scope than this PR.

File: silkworm/silkrpc/storage/remote_chain_storage.cpp [hunk 2]
```diff
@@ -54,8 +54,13 @@ Task<bool> RemoteChainStorage::read_block(const Hash& hash, BlockNum number, sil
...
 
 Task<bool> RemoteChainStorage::read_block(const Hash& hash, silkworm::Block& block) const {
-    const auto number = co_await core::rawdb::read_header_number(reader_, hash);
-    co_return co_await backend_->get_block(number, hash.bytes, /*.read_senders=*/false, block);
+    uint64_t block_number{0};
+    try {
+        block_number = co_await core::rawdb::read_header_number(reader_, hash);
+    } catch (const std::invalid_argument& iv) {
+        co_return false;
+    }
```
Comment (canepat) [inline silkworm/silkrpc/storage/remote_chain_storage.cpp:62]: As discussed, we should still rely on `std::invalid_argument` raised by low-level functions here without catching it at this level. Changing this behaviour consistently in the internal APIs is IMHO has broader scope than this PR.

**Review by canepat — APPROVED** [review]
These changes look very neat in the API implementation module, so I'm going to merge this as it is. Please consider to introduce the suggested changes in next PR

File: silkworm/silkrpc/core/cached_chain.cpp (hunk 1 above)
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.cpp:45]: OK

File: silkworm/silkrpc/core/cached_chain.cpp [hunk 3]
```diff
@@ -36,6 +36,32 @@ awaitable<std::shared_ptr<BlockWithHash>> read_block_by_number(BlockCache& cache
...
+    const auto cached_block = cache.get(block_hash);
+    if (cached_block) {
+        co_return cached_block.value();
+    }
+    auto block_with_hash = std::make_shared<BlockWithHash>();
+    auto block_found = co_await storage->read_block(block_hash, block_with_hash->block);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.cpp:51]: ok

File: silkworm/silkrpc/core/cached_chain.cpp [hunk 4]
```diff
@@ -36,6 +36,32 @@ awaitable<std::shared_ptr<BlockWithHash>> read_block_by_number(BlockCache& cache
...
+    if (block_found) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.cpp:52]: ok

File: silkworm/silkrpc/core/cached_chain.cpp [hunk 5]
```diff
@@ -50,6 +76,26 @@ awaitable<std::shared_ptr<BlockWithHash>> read_block_by_hash(BlockCache& cache,
     co_return block_with_hash;
 }
 
+awaitable<std::shared_ptr<BlockWithHash>> read_block_by_hash(BlockCache& cache, const std::shared_ptr<ChainStorage> storage, const evmc::bytes32& block_hash) {
+    const auto cached_block = cache.get(block_hash);
+    if (cached_block) {
+        co_return cached_block.value();
+    }
+    auto block_with_hash = std::make_shared<BlockWithHash>();
+    auto block_found = co_await storage->read_block(block_hash, block_with_hash->block);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.cpp:85]: ok

File: silkworm/silkrpc/core/cached_chain.cpp [hunk 6]
```diff
@@ -50,6 +76,26 @@ awaitable<std::shared_ptr<BlockWithHash>> read_block_by_hash(BlockCache& cache,
...
+    if (block_found) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.cpp:86]: ok

File: silkworm/silkrpc/core/cached_chain.hpp [hunk 7]
```diff
@@ -31,7 +32,9 @@ namespace silkworm::rpc::core {
 using boost::asio::awaitable;
 
 awaitable<std::shared_ptr<BlockWithHash>> read_block_by_number(BlockCache& cache, const rawdb::DatabaseReader& reader, uint64_t block_number);
+awaitable<std::shared_ptr<BlockWithHash>> read_block_by_number(BlockCache& cache, const std::shared_ptr<ChainStorage> storage, const rawdb::DatabaseReader& reader, uint64_t block_number);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.hpp:35]: ok

File: silkworm/silkrpc/core/cached_chain.hpp [hunk 8]
```diff
@@ -31,7 +32,9 @@ namespace silkworm::rpc::core {
...
 awaitable<std::shared_ptr<BlockWithHash>> read_block_by_hash(BlockCache& cache, const rawdb::DatabaseReader& reader, const evmc::bytes32& block_hash);
+awaitable<std::shared_ptr<BlockWithHash>> read_block_by_hash(BlockCache& cache, const std::shared_ptr<ChainStorage> storage, const evmc::bytes32& block_hash);
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/cached_chain.hpp:37]: ok

File: silkworm/silkrpc/storage/remote_chain_storage.cpp (hunk 2 above)
Comment (lupin012) [reply silkworm/silkrpc/storage/remote_chain_storage.cpp:62]: ok

**Pull request merged**